eth-rpc/client/clientdispatch.nim

```diff
@@ -77,37 +89,102 @@ proc connect*(self: RpcClient, address: string, port: Port) {.async.} =
   self.port = port
   asyncCheck processData(self)
 
-proc makeTemplate(name: string, params: NimNode, body: NimNode, starred: bool): NimNode =
-  # set up template AST
-  result = newNimNode(nnkTemplateDef)
-  if starred: result.add postFix(ident(name), "*")
-  else: result.add ident(name)
-  result.add newEmptyNode(), newEmptyNode(), params, newEmptyNode(), newEmptyNode(), body
-
-proc appendFormalParam(formalParams: NimNode, identName, typeName: string) =
-  # set up formal params AST
-  formalParams.expectKind(nnkFormalParams)
-  if formalParams.len == 0: formalParams.add newEmptyNode()
-  var identDef = newIdentDefs(ident(identName), ident(typeName))
-  formalParams.add identDef
-
-macro generateCalls: untyped =
-  ## Generate templates for client calls so that:
-  ##   client.call("web3_clientVersion", params)
-  ## can be written as:
-  ##   client.web3_clientVersion(params)
+import ../ jsonmarshal
+
+proc createRpcProc(procName, parameters, callBody: NimNode): NimNode =
+  # parameters come as a tree
+  var paramList = newSeq[NimNode]()
+  for p in parameters: paramList.add(p)
+
+  result = newProc(procName, paramList, callBody)           # build proc
+  result.addPragma ident"async"                             # make proc async               
+  result[0] = nnkPostFix.newTree(ident"*", newIdentNode($procName))  # export this proc
+
+proc toJsonArray(parameters: NimNode): NimNode =
+  # outputs an array of jsonified parameters
+  # ie; %[%a, %b, %c]
+  parameters.expectKind nnkFormalParams
+  var items = newNimNode(nnkBracket)
+  for i in 2 ..< parameters.len:
+    let curParam = parameters[i][0]
+    if curParam.kind != nnkEmpty:
+      items.add(nnkPrefix.newTree(ident"%", curParam))
+  result = nnkPrefix.newTree(newIdentNode("%"), items)
```

Comment:
You can try to use `bindSym("%", brForceOpen)` here instead of using `newIdentNode`.